File: src/main/java/fr/adrienbrault/idea/symfonyplugin/codeInspection/form/FormTypeAsClassConstantInspection.java
package fr.adrienbrault.idea.symfonyplugin.codeInspection.form;

import com.intellij.codeInspection.LocalInspectionTool;
import com.intellij.codeInspection.LocalQuickFixOnPsiElement;
import com.intellij.codeInspection.ProblemHighlightType;
import com.intellij.codeInspection.ProblemsHolder;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiElementVisitor;
import com.intellij.psi.PsiFile;
import com.intellij.util.ArrayUtil;
import com.jetbrains.php.lang.psi.elements.MethodReference;
import com.jetbrains.php.lang.psi.elements.StringLiteralExpression;
import fr.adrienbrault.idea.symfonyplugin.Symfony2ProjectComponent;
import fr.adrienbrault.idea.symfonyplugin.form.util.FormUtil;
import fr.adrienbrault.idea.symfonyplugin.util.PhpElementsUtil;
import fr.adrienbrault.idea.symfonyplugin.util.SymfonyUtil;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
public class FormTypeAsClassConstantInspection extends LocalInspectionTool {

    public static String MESSAGE = "Use fully-qualified class name (FQCN)";

    @NotNull
    public PsiElementVisitor buildVisitor(final @NotNull ProblemsHolder holder, boolean isOnTheFly) {
        if(!Symfony2ProjectComponent.isEnabled(holder.getProject()) || !SymfonyUtil.isVersionGreaterThenEquals(holder.getProject(), "2.8")) {
            return super.buildVisitor(holder, isOnTheFly);
        }

        return new MyPsiElementVisitor(holder);
    }

    private static class MyPsiElementVisitor extends PsiElementVisitor {
        private final ProblemsHolder holder;

        MyPsiElementVisitor(ProblemsHolder holder) {
            this.holder = holder;
        }

        @Override
        public void visitElement(PsiElement element) {
            if(!(element instanceof MethodReference) ||
                !ArrayUtil.contains(((MethodReference) element).getName(), "add", "create") ||
                !PhpElementsUtil.isMethodReferenceInstanceOf((MethodReference) element, "Symfony\\Component\\Form\\FormBuilderInterface")
                ) {

                super.visitElement(element);
                return;
            }

            PsiElement[] parameters = ((MethodReference) element).getParameters();
            if(parameters.length < 2) {
                super.visitElement(element);
                return;
            }

            if(!(parameters[1] instanceof StringLiteralExpression) || ((StringLiteralExpression) parameters[1]).getContents().contains("\\")) {
                super.visitElement(element);
                return;
            }

            holder.registerProblem(
                parameters[1],
                MESSAGE,
                ProblemHighlightType.WEAK_WARNING, new MyLocalQuickFix(parameters[1])
            );

            super.visitElement(element);
        }

        private static class MyLocalQuickFix extends LocalQuickFixOnPsiElement {

            MyLocalQuickFix(@NotNull PsiElement element) {
                super(element);
            }

            @NotNull
            @Override
            public String getText() {
                return "Use class constant";
            }

            @Nls
            @NotNull
            @Override
            public String getFamilyName() {
                return "Class constant";
            }

            @Override
            public void invoke(@NotNull Project project, @NotNull PsiFile psiFile, @NotNull PsiElement psiElement, @NotNull PsiElement psiElement1) {
                if(!(getStartElement() instanceof StringLiteralExpression)) {
                    return;
                }

                try {
                    FormUtil.replaceFormStringAliasWithClassConstant((StringLiteralExpression) getStartElement());
                } catch (Exception ignored) {
                }
            }
        }
    }
}

File: src/main/kotlin/fr/adrienbrault/idea/symfony2plugin/codeInspection/form/FormTypeAsClassConstantInspection.kt
package fr.adrienbrault.idea.symfony2plugin.codeInspection.form

import com.intellij.codeInspection.LocalInspectionTool
import com.intellij.codeInspection.LocalQuickFixOnPsiElement
import com.intellij.codeInspection.ProblemHighlightType
import com.intellij.codeInspection.ProblemsHolder
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiElementVisitor
import com.intellij.psi.PsiFile
import com.jetbrains.php.lang.psi.elements.MethodReference
import com.jetbrains.php.lang.psi.elements.StringLiteralExpression
import fr.adrienbrault.idea.symfony2plugin.Symfony2ProjectComponent
import fr.adrienbrault.idea.symfony2plugin.form.util.FormUtil
import fr.adrienbrault.idea.symfony2plugin.util.PhpElementsUtil
import fr.adrienbrault.idea.symfony2plugin.util.SymfonyUtil

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
open class FormTypeAsClassConstantInspection : LocalInspectionTool() {
    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
        if (!Symfony2ProjectComponent.isEnabled(holder.project)) {
            return super.buildVisitor(holder, isOnTheFly)
        }

        return MyPsiElementVisitor(holder)
    }

    private class MyPsiElementVisitor(private val holder: ProblemsHolder) : PsiElementVisitor() {
        private val isVersionGreaterThenEquals by lazy(LazyThreadSafetyMode.NONE) {
            SymfonyUtil.isVersionGreaterThenEquals(holder.project, "2.8")
        }

        override fun visitElement(element: PsiElement) {
            if (element !is MethodReference ||
                element.name != "add" && element.name != "create" ||
                !PhpElementsUtil.isMethodReferenceInstanceOf(element, "Symfony\\Component\\Form\\FormBuilderInterface")
            ) {
                super.visitElement(element)
                return
            }

            if (!isVersionGreaterThenEquals) {
                return
            }

            val parameters = element.parameters
            if (parameters.size < 2) {
                super.visitElement(element)
                return
            }

            val formType = parameters[1]
            if (formType !is StringLiteralExpression || "\\" in formType.contents) {
                super.visitElement(element)
                return
            }

            holder.registerProblem(
                formType,
                MESSAGE,
                ProblemHighlightType.WEAK_WARNING,
                MyLocalQuickFix(formType)
            )

            super.visitElement(element)
        }

        private class MyLocalQuickFix(element: PsiElement) : LocalQuickFixOnPsiElement(element) {
            override fun getText() = "Use class constant"

            override fun getFamilyName() = "Class constant"

            override fun invoke(project: Project, psiFile: PsiFile, psiElement: PsiElement, psiElement1: PsiElement) {
                val formType = startElement as? StringLiteralExpression ?: return

                try {
                    FormUtil.replaceFormStringAliasWithClassConstant(formType)
                } catch (_: Exception) {
                }
            }
        }
    }

    companion object {
        const val MESSAGE = "Use fully-qualified class name (FQCN)"
    }
}
